### PR TITLE
Azure AD Access Token support for DatabaseTarget 

### PR DESCRIPTION
**Pull request by 304NotModified**
Based on #3824

fixes https://github.com/NLog/NLog/issues/3554

### CODE REVIEW

**Comment by snakefoot** [comment]
Think Connection-Properties should be dynamic just like Command-Parameters.

Instead of DBAccessToken then it should be a list of connetion-property-items. Each with a property-name and property-value-layout.

**Comment by 304NotModified** [comment]
That's indeed also an option, but a bit more work

**Comment by snakefoot** [comment]
> That's indeed also an option, but a bit more work

Well elegant generic solutions requires work.

Maybe your first suggestion of creating an MsSqlConnection-specific- uget-package is better. Maybe based on SqlBulkCopy.

**Comment by anxkha** [comment]
@304NotModified Is there an easy way to get at a preview build of this on nuget so I can test drive it and put it in my project? Or only for the next RC of 4.7.0 or whatever comes next?

**Comment by 304NotModified** [comment]
> @304NotModified Is there an easy way to get at a preview build of this on nuget so I can test drive it and put it in my project? Or only for the next RC of 4.7.0 or whatever comes next?

Yes, it will be in the appveyor output when the build is done.

**Comment by snakefoot** [comment]
Since you are walking down the path of creating yet another dynamic-setter-logic, then I think it is time to extract the logic from `class DbTypeSetter` into `class PropertyHelper` and reuse this logic for assigning DbConnection-properties.

**Comment by 304NotModified** [comment]
good idea for NLog 5. I will try not to refactor too much for NLog 4.x

**Comment by snakefoot** [comment]
Adding new features are not refactoring.

**Comment by 304NotModified** [comment]
> Since you are walking down the path of creating yet another dynamic-setter-logic, then I think it is time to extract the logic from class DbTypeSetter into class PropertyHelper

This is refactoring?

**Comment by snakefoot** [comment]
It is only this logic you should extract:

`PropertyInfo propInfo = dbParameterType.GetProperty(dbTypeNames[0], BindingFlags.IgnoreCase | BindingFlags.Public | BindingFlags.Instance);`

And this logic:

```c#
var dbTypeSetterLambda = ReflectionHelpers.CreateLateBoundMethod(_dbTypeSetter.GetSetMethod());
var dbTypeSetterParams = new object[] { _dbTypeValue };
_dbTypeSetterFast = (p) => dbTypeSetterLambda.Invoke(p, dbTypeSetterParams);
```

**Comment by 304NotModified** [comment]
I could create a more generic approach of AccessTokenSetter, e.g. PropertySetter. That will make this easier to set other properties on the Connection. What do you think?

**Comment by snakefoot** [comment]
> I could create a more generic approach of AccessTokenSetter, e.g. PropertySetter

Sound good. If you can make a single ConnectionPropertySetter then it is super easy to iterate a list of them (Item with property-name + value-layout)

**Comment by 304NotModified** [comment]
there is a difference between 

DbTypeSetter: 
    _dbTypeSetterFast = (p) => dbTypeSetterLambda.Invoke(p, dbTypeSetterParams);

and the "setterFast" for this will have 2 parameters. I don't see how to combine those different approaches. Any suggestions?

**Comment by snakefoot** [comment]
>  Any suggestions?

Of course the generic version should allow you to assign a dynamic-value using parameter-array. It is just `class DbTypeSetter` that knows the dynamic-value never changes and wraps the dynamic-value-setter in a lambda that captures the "static" array for reuse.

One could make an property-assigner-class that takes a value generator-func, and reuses the same parameter-array for each call (updating the first parameter before calling). Ofcourse it is restricted so it takes lock on the internal parameter-array while calling the setter-method.

**Comment by snakefoot** [comment]
Actually the logic about reusing an internal parameter-array should just be placed in the value-generator-func

**Comment by 304NotModified** [comment]
I've create the PropertySetter, but haven't refactored the DbTypeSetter yet. Let me know what you think!

**Comment by snakefoot** [comment]
You just need to 2 things:

- Add a `SetValue`-method that takes a pre-allocated parameter-array (object[]),
- Add new method for calling `ReflectionHelpers.CreateLateBoundMethod(_propertySetter);

Then it can be used by DbTypeSetter.

**Comment by snakefoot** [comment]
If you are going to use it for AccessToken, then I guess you need to make an interface for PropertySetter called IPropertySetter. And create the propertysetter like this:

```
var propertySetter = (IPropertySetter)Activator.CreateInstance(typeof(PropertySetter<>).MakeGenericType(propInfo.PropertyType));
```

Alternative change `PropertySetter` to be non-generic and only get object-type and property-name.

**Comment by 304NotModified** [comment]
Or for non generic approach, pass object as T? I could even makes T optional (defaults to object)

**Comment by 304NotModified** [comment]
> Add a SetValue-method that takes a pre-allocated parameter-array

And then reuse the array every time?

**Comment by snakefoot** [comment]
> Or for non generic approach, pass object as T? I could even makes T optional (defaults to object)

Right now there only two cases of types. Custom DbType and String. For both cases then the object-value will do just fine.

But I guess some day in the feature somebody wants to assign an integer (Either for the DbConnection, or maybe one day for the DbCommand). And then the ObjectSetter needs to inspet the PropertyType and attempt conversion from input-value-type (possible string or rawvalue) to the actual type.

**Comment by snakefoot** [comment]
> > Add a SetValue-method that takes a pre-allocated parameter-array

> And then reuse the array every time?

Nope it is the caller that has to decide whether to reuse the same array every time. For DbType it is done by doing delegate-capture of the parameter-array.

File: src/NLog/Internal/PropertySetter.cs [hunk 1]
```diff
@@ -60,8 +75,50 @@ private PropertySetter([NotNull] PropertyInfo property)
...
+    /// </summary>
+    /// <typeparam name="T">The type of the property. Not used for detection, but only for compile-time safe <see cref="SetValue"/></typeparam>
+    internal class PropertySetter<T>
+    {
+        private readonly ReflectionHelpers.LateBoundMethod _propertySetter;
+
+        /// <summary>
+        /// array with 1 item. The property will always have one property
+        /// </summary>
+        private readonly object[] _arrayWith1Item = new object[1];
```
Comment (snakefoot) [inline src/NLog/Internal/PropertySetter.cs]: Having the object-reuse here will give race-condition issues for those not aware. I would move this object-reuse out of the PropertySetter-class (Have SetValue-method that takes array)

Comment (snakefoot) [reply src/NLog/Internal/PropertySetter.cs]: The array-reuse is mostly important for custom DbType. For Connection-properties then a single-item-array allocation is nothing compare to the allocation performed by DbCommand.

**Comment by 304NotModified** [comment]
I like to do what you're asking, but it too hard and I don't want to burn my fingers on the DbTypeSetter . I think this good for now and I'm OK with refactoring this further in the future.

File: src/NLog/Internal/PropertySetter.cs (hunk 1 above)
Comment (304NotModified) [reply src/NLog/Internal/PropertySetter.cs]: I think we should fix the array thing outside this PR. I think the latebound creator could be improved. Will create a new PR for that.

**Comment by snakefoot** [comment]
Created #3829 to skip the journey with future refactoring and trying to mark DBAccessToken as obsolete (With lots of battles because the property has just been added, and it doesn't look good to mark it obsolete so quickly)

**Comment by 304NotModified** [comment]
Ok cool!

 I'm on a short holiday from tomorrow. Im back March 11 and will continue this and release 4.7 shortly. (I tried finishing 4.7 before my holiday ;) but quality of quantity :+1:)

**Comment by 304NotModified** [comment]
superseded by #3829

**Pull request closed without merging**